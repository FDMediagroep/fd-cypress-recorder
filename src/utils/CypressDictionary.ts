export enum FdEventType {
    ATTRIBUTE_VALUE_CONTAINS = 'attribute-value-contains',
    ATTRIBUTE_VALUE_EQUALS = 'attribute-value-equals',
    ATTRIBUTE_VALUE_EXISTS = 'attribute-value-exists',
    CLEAR_COOKIES = 'clear-cookies',
    CLICK = 'click',
    CONTAINS_TEXT = 'contains-text',
    HOVER = 'hover',
    LOCATION = 'location',
    LOCATION_CONTAINS = 'location-contains',
    EXISTS = 'exists',
    TYPE = 'type',
    VIEWPORT_SIZE = 'viewport-size',
    VISIT = 'visit',
}

export interface FdEvent {
    type: FdEventType;
}

export interface FdAttributeValueEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
    value: string;
}

export interface FdAttributeExistsEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
}

export interface FdTypeEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdClickEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdHoverEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdLocationEvent extends FdEvent {
    href: string;
}

export interface FdLocationContainsEvent extends FdEvent {
    value: string;
}

export interface FdVisitEvent extends FdEvent {
    href: string;
}

export interface FdTextContentEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdViewportSizeEvent extends FdEvent {
    width: number;
    height: number;
}

export interface FdExistsEvent extends FdEvent {
    target: string; // CSS Selector
}

export type AllFdEvents = FdEvent
    | FdAttributeValueEvent
    | FdAttributeExistsEvent
    | FdClickEvent
    | FdHoverEvent
    | FdLocationEvent
    | FdLocationContainsEvent
    | FdVisitEvent
    | FdTextContentEvent
    | FdViewportSizeEvent
    | FdExistsEvent
    | FdTypeEvent;

export interface Template {
    name: string;
    description?: string;
    events: AllFdEvents[];
}

export const UNIQUE_SELECTOR_OPTIONS = {
    selectorTypes: ['Tag', 'NthChild']
};

export interface Options {
    basicAuth?: boolean;
}

export function getCode(event: AllFdEvents, options?: Options) {
    switch (event.type) {
        case FdEventType.CLEAR_COOKIES:
            return `cy.clearCookies();`;
        case FdEventType.CLICK:
            return `cy.get('${(event as FdClickEvent).target}').should('exist').click();`;
        case FdEventType.CONTAINS_TEXT:
            return `cy.get('${(event as FdTextContentEvent).target}').contains('${(event as FdTextContentEvent).value.replace(new RegExp("'", 'g'), "\\\'")}');`;
        case FdEventType.EXISTS:
            return `cy.get('${(event as FdExistsEvent).target}').should('exist');`;
        case FdEventType.HOVER:
            return `cy.get('${(event as FdHoverEvent).target}').trigger('mouseover');`;
        case FdEventType.LOCATION:
            return `cy.location('href', {timeout: 10000}).should('eq', '${(event as FdLocationEvent).href}');`;
        case FdEventType.LOCATION_CONTAINS:
            return `cy.location('href', {timeout: 10000}).should('contain', '${(event as FdLocationContainsEvent).value}');`;
        case FdEventType.VISIT:
            if (options && options.basicAuth) {
                return `cy.visit('${(event as FdLocationEvent).href}', {auth: {username: Cypress.env('BASIC_USER') || '', password: Cypress.env('BASIC_PASS') || ''}});`;
            } else {
                return `cy.visit('${(event as FdLocationEvent).href}');`;
            }
        case FdEventType.TYPE:
            return `cy.get('${(event as FdTypeEvent).target}').type('${(event as FdTypeEvent).value}');`;
        case FdEventType.VIEWPORT_SIZE:
            return `cy.viewport(${(event as FdViewportSizeEvent).width}, ${(event as FdViewportSizeEvent).height});`;
        case FdEventType.ATTRIBUTE_VALUE_EQUALS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el: any) => el.attr('${(event as FdAttributeValueEvent).name}')).then((attr: string) => expect(attr).to.eq('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'))`;
        case FdEventType.ATTRIBUTE_VALUE_CONTAINS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el: any) => el.attr('${(event as FdAttributeValueEvent).name}')).then((attr: string) => expect(attr).to.contain('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'))`;
        case FdEventType.ATTRIBUTE_VALUE_EXISTS:
            return `cy.get('${(event as FdAttributeExistsEvent).target}').should('have.attr', '${(event as FdAttributeExistsEvent).name}');`;
        default:
            return `// ${JSON.stringify(event)}`;
    }
}
