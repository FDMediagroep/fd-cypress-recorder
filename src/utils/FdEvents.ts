export interface Header {
    property: string;
    value: string;
}

export enum FdEventType {
    ATTRIBUTE_VALUE_CONTAINS = 'attribute-value-contains',
    ATTRIBUTE_VALUE_EQUALS = 'attribute-value-equals',
    ATTRIBUTE_VALUE_EXISTS = 'attribute-value-exists',
    CLEAR_COOKIES = 'clear-cookies',
    CLICK = 'click',
    CONTAINS_TEXT = 'contains-text',
    COUNT_EQUALS = 'count-equals',
    COUNT_GREATER_THAN = 'count-greater-than',
    COUNT_LESS_THAN = 'count-less-than',
    HOVER = 'hover',
    LOCATION = 'location',
    LOCATION_CONTAINS = 'location-contains',
    EXISTS = 'exists',
    /**
     * Type text
     */
    TYPE = 'type',
    VIEWPORT_SIZE = 'viewport-size',
    VISIT = 'visit',
}

export interface FdEvent {
    type: FdEventType;
}

export interface FdAttributeValueEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
    value: string;
}

export interface FdAttributeExistsEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
}

export interface FdTypeEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdClickEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdHoverEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdLocationEvent extends FdEvent {
    href: string;
}

export interface FdLocationContainsEvent extends FdEvent {
    value: string;
}

export interface FdVisitEvent extends FdEvent {
    href: string;
}

export interface FdTextContentEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdViewportSizeEvent extends FdEvent {
    width: number;
    height: number;
}

export interface FdExistsEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdCountEqualsEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export interface FdCountGreaterThanEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export interface FdCountLessThanEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export type AllFdEvents = FdEvent
    | FdAttributeValueEvent
    | FdAttributeExistsEvent
    | FdClickEvent
    | FdCountEqualsEvent
    | FdCountGreaterThanEvent
    | FdCountLessThanEvent
    | FdExistsEvent
    | FdHoverEvent
    | FdLocationEvent
    | FdLocationContainsEvent
    | FdTextContentEvent
    | FdTypeEvent
    | FdViewportSizeEvent
    | FdVisitEvent
;

export interface Template {
    name: string;
    description?: string;
    events: AllFdEvents[];
}

export const UNIQUE_SELECTOR_OPTIONS = {
    selectorTypes: ['Tag', 'NthChild']
};

export interface Options {
    basicAuth?: boolean;
    headers?: Header[];
}
