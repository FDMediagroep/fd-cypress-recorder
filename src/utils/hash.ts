declare let Uint8Array: Uint8ArrayConstructor;
/**
 * Convert given ArrayBuffer to HEX-string.
 * @param buffer
 */
export function hexString(buffer: ArrayBuffer) {
    const byteArray = new Uint8Array(buffer);

    const hexCodes = [...byteArray].map((value) => {
        const hexCode = value.toString(16);
        const paddedHexCode = hexCode.padStart(2, '0');
        return paddedHexCode;
    });

    return hexCodes.join('');
}

/**
 * Digest the given message. ArrayBuffer result is returned in a Promise.
 * @param message
 */
export function digestMessage(message: string) {
    const encoder = new TextEncoder();
    const data = encoder.encode(message);
    return window.crypto.subtle.digest('SHA-256', data);
}

/**
 * Returns digest of given message in the form of a HEX-string.
 * @param message
 */
export async function digest(message: string) {
    const arrayBuffer = await digestMessage(message);
    const hex = hexString(arrayBuffer);
    return hex;
}

/**
 * Returns true when the sum of the ArrayBuffer digest of given message is even and false when odd.
 * @param message
 */
export async function isEven(message: string) {
    const arrayBuffer = await digestMessage(message);
    const byteArray = new Uint8Array(arrayBuffer);
    return (
        byteArray.reduce(
            (previousValue: number, currentValue: number) =>
                previousValue + currentValue
        ) %
            2 ===
        0
    );
}
