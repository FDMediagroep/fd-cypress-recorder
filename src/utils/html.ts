/**
 * Embed given string as JavaScript in the user-DOM space.
 * @param fn
 */
export function embed(fn: (text: string) => void) {
    const script = document.createElement("script");
    script.text = `(${fn.toString()})();`;
    document.documentElement.appendChild(script);
}

/**
 * Convert the given HTML-string to a DOM-node.
 * @param html
 */
export function htmlToElement(html: string): ChildNode {
    const template = document.createElement('template');
    html = html.trim(); // Never return a text node of whitespace as the result
    template.innerHTML = html;
    return template.content.firstChild as ChildNode;
}
