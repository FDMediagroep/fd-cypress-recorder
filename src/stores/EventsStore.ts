import { StoreBase, AutoSubscribeStore, autoSubscribe } from 'resub';
import { AllFdEvents } from '../utils/FdEvents';

@AutoSubscribeStore
class EventsStore extends StoreBase {
    private events: AllFdEvents[] = [];
    private undoneEvents: AllFdEvents[] = [];
    private futures: any = [];

    stepBack() {
        const currentEvents = [...this.events];
        if (currentEvents.length) {
            currentEvents.pop();
            this.events = currentEvents;
            this.trigger('stepBack');
        }
    }

    addFuture(events: AllFdEvents[]) {
        this.futures = [...this.futures, events];
    }

    popFuture() {
        let future;
        const futures = [...this.futures];
        if (futures.length) {
            future = futures.pop();
        }
        this.futures = futures;
        return future;
    }

    clearUndone() {
        this.undoneEvents = [];
    }

    addEvent(event: AllFdEvents) {
        this.clearUndone();
        this.events = this.events.concat(event);
        this.trigger('addEvent');
    }

    /**
     * Add event if it does not already exist.
     */
    addUniqueEvent(event: AllFdEvents) {
        const foundEvent = this.events.filter((ev) => ev.type === event.type);
        if (foundEvent.length === 0) {
            this.clearUndone();
            this.events = [...this.events, event];
        }
        this.trigger('addUniqueEvent');
    }

    /**
     * Add event if not already exists. Otherwise override existing.
     */
    setEvent(event: AllFdEvents) {
        const foundEvent = this.events.filter((ev) => ev.type === event.type);
        this.clearUndone();
        if (foundEvent.length === 0) {
            this.events = this.events.concat(event);
        } else {
            const copy = [...this.events];
            copy.forEach((ev, idx) => {
                if (ev.type === event.type) {
                    copy[idx] = event;
                }
            });
            this.events = copy;
        }
        this.trigger('setEvent');
    }

    setEvents(
        events: AllFdEvents[],
        triggerKey?: string | number | string[] | number[]
    ) {
        if (this.events && events && this.events.length !== events.length) {
            this.clearUndone();
        }
        this.events = events || [];
        this.trigger(triggerKey ? triggerKey : 'setEvents');
    }

    clear() {
        this.clearUndone();
        this.events = [];
        this.trigger('clear');
    }

    @autoSubscribe
    getEvents() {
        return this.events;
    }

    @autoSubscribe
    getUndoneEvents() {
        return this.undoneEvents;
    }

    @autoSubscribe
    getFutures() {
        return this.futures;
    }
}

export = new EventsStore();
