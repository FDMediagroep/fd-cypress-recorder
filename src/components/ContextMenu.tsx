import React, { PureComponent } from "react";
import styled from "styled-components";
import EventsStore = require("../stores/EventsStore");
import { FdEventType, FdAttributeValueEvent, FdAttributeExistsEvent, AllFdEvents } from "../utils/CypressDictionary";
import ContextULCheckAttribute from "./ContextULCheckAttribute";
import ContextULCheckCount from "./ContextULCheckCount";

declare var window: Window;

export interface Props {
    target: HTMLElement;
    selector: string;
}

/**
 * This component renders the context menu.
 */
export default class ContextMenu extends PureComponent<Props, any> {
    state: any = {};
    private top = 0;
    private left = 0;
    private w = 0;

    constructor(props: Props) {
        super(props);
        this.top = Math.max(document.documentElement.scrollTop + this.props.target.getBoundingClientRect().top + this.props.target.getBoundingClientRect().height, 0);
        this.left = Math.max(document.documentElement.scrollLeft + this.props.target.getBoundingClientRect().left + this.props.target.getBoundingClientRect().width, 0);
        this.w = Math.max(document.documentElement.clientWidth, window.innerWidth || 0);
        if (this.left + 250 > this.w) {
            this.left = Math.max(document.documentElement.scrollLeft + this.props.target.getBoundingClientRect().left - 200, 0);
        }
    }

    handleClick = () => {
        EventsStore.addEvent({ type: FdEventType.CLICK, target: this.props.selector });
        this.props.target.click();
    }

    handleCheckExists = () => {
        EventsStore.addEvent({ type: FdEventType.EXISTS, target: this.props.selector });
    }

    handleCheckText = () => {
        const textContent = prompt('Text content of selected element should contain') || '';
        if (textContent) {
            EventsStore.addEvent({ type: FdEventType.CONTAINS_TEXT, target: this.props.selector, value: textContent });
            alert(`Check text is "${textContent}" added`);
        }
    }

    handleHover = () => {
        EventsStore.addEvent({ type: FdEventType.HOVER, target: this.props.selector });
        this.props.target.focus();
    }

    handleVisit = () => {
        EventsStore.addEvent({ type: FdEventType.VISIT, href: window.location.href });
    }

    handleAwaitLocation = () => {
        EventsStore.addEvent({ type: FdEventType.LOCATION, href: window.location.href });
    }

    handleAwaitLocationContains = () => {
        const value = prompt('URL should contain') || '';
        if (value) {
            EventsStore.addEvent({ type: FdEventType.LOCATION_CONTAINS, value });
        }
    }

    handleGoToLocation = () => {
        const value = prompt('URL to visit', 'https://') || '';
        if (value) {
            EventsStore.addEvent({ type: FdEventType.VISIT, href: value });
            window.location.assign(value);
        }
    }

    handleEnterText = () => {
        this.props.target.focus();
        const value = prompt('Type your text') || '';
        if (value) {
            EventsStore.addEvent({ type: FdEventType.TYPE, target: this.props.selector, value });
        }
    }

    handleSubContextMenuAssert = (event: AllFdEvents) => {
        EventsStore.addEvent(event);
    }

    /**
     * Open Check Attribute context menu
     */
    handleCheckAttribute = (e: React.MouseEvent<HTMLElement>) => {
        e.preventDefault();
        this.setState({
            customContextMenu: <ContextULCheckAttribute target={this.props.target} selector={this.props.selector} onMouseDown={this.handleSubContextMenuAssert} onBack={this.handleContextMenuBack}/>
        });
    }

    /**
     * Open Check Count context menu
     */
    handleCheckCount = (e: React.MouseEvent<HTMLElement>) => {
        e.preventDefault();
        this.setState({
            customContextMenu: <ContextULCheckCount target={this.props.target} selector={this.props.selector} onMouseDown={this.handleSubContextMenuAssert} onBack={this.handleContextMenuBack}/>
        });
    }

    handleContextMenuBack = () => {
        this.setState({customContextMenu: null});
    }

    render() {
        return (
            <StyledContextMenu top={this.top} left={this.left}>
                {this.state.customContextMenu ? this.state.customContextMenu :
                (
                    <ul>
                        <li className="label">Interactions</li>
                        <li className="clickable" onMouseDown={this.handleClick}>Click</li>
                        <li className="clickable" onMouseDown={this.handleEnterText}>Enter text...</li>
                        <li className="clickable" onMouseDown={this.handleHover}>Hover</li>
                        <li className="label">Asserts</li>
                        <li className="clickable" onMouseDown={this.handleCheckAttribute}>Attributes...</li>
                        <li className="clickable" onMouseDown={this.handleCheckText}>Contains text...</li>
                        <li className="clickable" onMouseDown={this.handleCheckCount}>Count...</li>
                        <li className="clickable" onMouseDown={this.handleCheckExists}>Exists</li>
                        <li className="label">Global</li>
                        <li className="clickable" onMouseDown={this.handleGoToLocation}>Go to URL...</li>
                        <li className="clickable" onMouseDown={this.handleAwaitLocation}>Match current URL</li>
                        <li className="clickable" onMouseDown={this.handleAwaitLocationContains}>URL contains...</li>
                        <li className="clickable" onMouseDown={this.handleVisit}>Visit current URL</li>
                    </ul>
                )}
                <small>{this.props.selector}</small>
            </StyledContextMenu>
        );
    }
}

const StyledContextMenu: any = styled.div`
    z-index: 2;
    position: absolute;
    background-color: #fff;
    border-radius: 5px;
    box-shadow: 0 0 5px #999;
    top: ${(props: any) => props.top}px;
    left: ${(props: any) => props.left}px;
    width: 250px;
    max-height: 500px;
    word-break: break-all;
    overflow: auto;
    overscroll-behavior: contain;
    ::-webkit-scrollbar {
        width: .5rem;
        height: .5rem;
    }

    ::-webkit-scrollbar-track {
        -webkit-border-radius: .25rem;
        border-radius: .25rem;
        background:rgba(0,0,0,0.1);
    }

    ::-webkit-scrollbar-thumb {
        -webkit-border-radius: .25rem;
        border-radius: .25rem;
        background:rgba(0,0,0,0.2);
    }

    ::-webkit-scrollbar-thumb:hover {
        background: rgba(0,0,0,0.4);
    }

    ::-webkit-scrollbar-thumb:window-inactive {
        background: rgba(0,0,0,0.05);
    }

    ul {
        list-style-type: none;
        padding: 0;
        margin: 0;
        li {
            display: flex;
            align-items: center;
            border-bottom: 1px solid rgba(0, 0, 0, 0.1);
            padding: .5rem 1rem;
            &.label {
                justify-content: center;
                color: #49a4a2;
                &.separator {
                    background-color: rgba(0, 0, 0, .1);
                }
                &.back {
                    cursor: pointer;
                }
            }
            &.clickable {
                cursor: pointer;
                &:hover {
                    text-decoration: underline;
                }
            }
        }
    }
    > small {
        color: #677381;
        display: block;
        padding: 1rem;
    }
`;
