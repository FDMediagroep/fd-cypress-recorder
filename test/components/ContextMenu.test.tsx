import React from 'react';
import {act} from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import ContextMenu from '../../src/components/ContextMenu';
import ReactDOM from 'react-dom';
import EventsStore = require('../../src/stores/EventsStore');
import { FdEventType, FdClickEvent, FdVisitEvent, FdHoverEvent, FdExistsEvent, FdLocationEvent, FdTextContentEvent, FdTypeEvent, FdLocationContainsEvent } from '../../src/utils/FdEvents';

describe('Context Menu', () => {
    let target: HTMLElement;
    const clickMock = jest.fn();
    const mouseDownEvt = new MouseEvent('mousedown', {bubbles: true});

    beforeEach(() => {
        target = document.createElement('div');
        target.setAttribute('class', 'context-menu-test');
        target.addEventListener('click', clickMock);
        document.body.appendChild(target);
        act(() => {
            ReactDOM.render(<ContextMenu target={target} selector={'.context-menu-test'}/>, target);
        });
        
    });

    afterEach(() => {
        target.removeEventListener('click', clickMock);
        document.body.removeChild(target);
        target.remove();
        EventsStore.clear();
    });

    it('should render context menu correctly', () => {
        let contextMenu = TestRenderer.create(<ContextMenu target={target} selector={'.context-menu-test'}/>);
        expect(contextMenu.toJSON()).toMatchSnapshot();
    });

    it('should handle manual Click event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'click') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdClickEvt = events[0] as FdClickEvent;
        expect(fdClickEvt.type).toBe(FdEventType.CLICK);
        expect(fdClickEvt.target).toBe('.context-menu-test');
    });

    it('should handle manual Text entering event correctly', () => {
        const promptMock  = jest.fn();
        promptMock.mockReturnValue('typing some text');
        window.prompt = promptMock;
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'enter text...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdEvt = events[0] as FdTypeEvent;
        expect(fdEvt.type).toBe(FdEventType.TYPE);
        expect(fdEvt.value).toBe('typing some text');
    });

    it('should handle manual Contains text event correctly', () => {
        const promptMock  = jest.fn();
        promptMock.mockReturnValue('some random text');
        window.prompt = promptMock;
        window.location.assign = jest.fn();
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'contains text...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdEvt = events[0] as FdTextContentEvent;
        expect(fdEvt.type).toBe(FdEventType.CONTAINS_TEXT);
        expect(fdEvt.value).toBe('some random text');
    });

    it('should handle Check Exists event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'exists') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdClickEvt = events[0] as FdExistsEvent;
        expect(fdClickEvt.type).toBe(FdEventType.EXISTS);
        expect(fdClickEvt.target).toBe('.context-menu-test');
    });

    it('should handle Hover event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'hover') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdClickEvt = events[0] as FdHoverEvent;
        expect(fdClickEvt.type).toBe(FdEventType.HOVER);
        expect(fdClickEvt.target).toBe('.context-menu-test');
    });

    it('should handle manual Go to URL event correctly', () => {
        const promptMock  = jest.fn();
        promptMock.mockReturnValue('http://willemliu.nl');
        window.prompt = promptMock;
        window.location.assign = jest.fn();
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'go to url...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdEvt = events[0] as FdVisitEvent;
        expect(fdEvt.type).toBe(FdEventType.VISIT);
        expect(fdEvt.href).toBe('http://willemliu.nl');
    });

    it('should handle Visit event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'visit current url') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdClickEvt = events[0] as FdVisitEvent;
        expect(fdClickEvt.type).toBe(FdEventType.VISIT);
        expect(fdClickEvt.href).toBe('http://localhost/');
    });

    it('should handle Match Current URL event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'match current url') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdClickEvt = events[0] as FdLocationEvent;
        expect(fdClickEvt.type).toBe(FdEventType.LOCATION);
        expect(fdClickEvt.href).toBe('http://localhost/');
    });

    it('should handle manual URL contains event correctly', () => {
        const promptMock  = jest.fn();
        promptMock.mockReturnValue('willemliu');
        window.prompt = promptMock;
        window.location.assign = jest.fn();
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'url contains...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        const events = EventsStore.getEvents();
        expect(events.length).toBe(1);
        const fdEvt = events[0] as FdLocationContainsEvent;
        expect(fdEvt.type).toBe(FdEventType.LOCATION_CONTAINS);
        expect(fdEvt.value).toBe('willemliu');
    });

    it('should handle Attributes... event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'attributes...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        expect(target.querySelectorAll('li.label.back').length).toBe(1);
    });

    it('should handle Attributes... and then Back event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'attributes...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        expect(target.querySelectorAll('li.label.back').length).toBe(1);

        [].slice.call(target.querySelectorAll('li.label.back')).forEach((li: HTMLLIElement) => {
            li.dispatchEvent(mouseDownEvt);
        });

        expect(target.querySelectorAll('li.label.back').length).toBe(0);
    });

    it('should handle Count... event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'count...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        expect(target.querySelectorAll('li.label.back').length).toBe(1);
    });

    it('should handle Count... and then Back event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'count...') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(clickMock).toHaveBeenCalledTimes(1);
        expect(target.querySelectorAll('li.label.back').length).toBe(1);

        [].slice.call(target.querySelectorAll('li.label.back')).forEach((li: HTMLLIElement) => {
            li.dispatchEvent(mouseDownEvt);
        });

        expect(target.querySelectorAll('li.label.back').length).toBe(0);
    });

});
