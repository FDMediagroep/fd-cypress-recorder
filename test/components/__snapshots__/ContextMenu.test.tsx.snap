// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Menu should render context menu correctly 1`] = `
<div
  className="sc-bwzfXH jxhsCW"
>
  <ul>
    <li
      className="label"
    >
      Interactions
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Click
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Hover
    </li>
    <li
      className="label"
    >
      Asserts
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Exists
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Contains text
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Attributes...
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      URL contains
    </li>
    <li
      className="label"
    >
      Global
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Match current URL
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Visit current URL
    </li>
    <li
      className="clickable"
      onMouseDown={[Function]}
    >
      Enter text
    </li>
  </ul>
  <small>
    .context-menu-test
  </small>
</div>
`;
